src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/writers/CreateEventGroup.kt

```diff
@@ -0,0 +1,70 @@
+// Copyright 2020 The Cross-Media Measurement Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.wfanet.measurement.kingdom.deploy.gcloud.spanner.writers
+
+import com.google.cloud.spanner.Value
+import org.wfanet.measurement.common.identity.ExternalId
+import org.wfanet.measurement.gcloud.spanner.bufferTo
+import org.wfanet.measurement.gcloud.spanner.insertMutation
+import org.wfanet.measurement.gcloud.spanner.set
+import org.wfanet.measurement.internal.kingdom.EventGroup
+import org.wfanet.measurement.kingdom.deploy.gcloud.spanner.common.KingdomInternalException
+import org.wfanet.measurement.kingdom.deploy.gcloud.spanner.readers.DataProviderReader
+import org.wfanet.measurement.kingdom.deploy.gcloud.spanner.readers.MeasurementConsumerReader
+
+/**
+ * @throws KingdomInternalException if the MeasurementConsumer or DataProvider for this EventGroup
+ * is not found.
+ */
+class CreateEventGroup(private val eventGroup: EventGroup) :
+  SpannerWriter<EventGroup, EventGroup>() {
+  override suspend fun TransactionScope.runTransaction(): EventGroup {
+    val measurementConsumerId =
+      MeasurementConsumerReader()
+        .readExternalIdOrNull(
+          transactionContext,
+          ExternalId(eventGroup.externalMeasurementConsumerId)
+        )
+        ?.measurementConsumerId
+        ?: throw KingdomInternalException(
+          KingdomInternalException.Code.MEASUREMENT_CONSUMER_NOT_FOUND
+        )
+
+    val dataProviderId =
+      DataProviderReader()
+        .readExternalIdOrNull(transactionContext, ExternalId(eventGroup.externalDataProviderId))
+        ?.dataProviderId
+        ?: throw KingdomInternalException(KingdomInternalException.Code.DATA_PROVIDER_NOT_FOUND)
+
+    val internalEventGroupId = idGenerator.generateInternalId()
+    val externalEventGroupId = idGenerator.generateExternalId()
+
+    insertMutation("EventGroups") {
+        set("EventGroupId" to internalEventGroupId.value)
+        set("ExternalEventGroupId" to externalEventGroupId.value)
+        set("MeasurementConsumerId" to measurementConsumerId)
+        set("DataProviderId" to dataProviderId)
+        set("ProvidedEventGroupId" to eventGroup.providedEventGroupId)
```

Comment:
The ProvidedEventGroupId is used for idempotency.

so somewhere it needs to check if an eventGroup with the same ProvidedEventGroupId exist or not. If yes, no need to create a new one, just return the existing one.